ambari-server/src/main/java/org/apache/ambari/server/checks/UpgradeCheckRegistry.java

```diff
@@ -17,165 +17,185 @@
+ * <li>Built-in: checks which ship with Ambari
+ * <li>Plug-in: checks which are specified in the upgrade pack for a stack, and
+ * either come from Ambari's checks or from the library classpath of the stack.
+ * </ul>
  */
-@Singleton
 public class UpgradeCheckRegistry {
   private static final Logger LOG = LoggerFactory.getLogger(UpgradeCheckRegistry.class);
 
+  @Inject
+  private Provider<AmbariMetaInfo> metainfoProvider;
+
   /**
-   * The list of upgrade checks to run through.
+   * Used for injecting dependencies into plugin upgrade check classes. There
+   * are cases where an upgrade pack specifies a built-in check which uses a
+   * dependency, such as the Ambari Configuration singleton.
    */
-  private Set<UpgradeCheck> m_upgradeChecks = new TreeSet<>(
+  @Inject
+  private Injector m_injector;
+
+  /**
+   * The list of upgrade checks provided by the system which are required, as
+   * specified by {@link UpgradeCheckInfo#required()}.
+   */
+  private final Set<UpgradeCheck> m_builtInChecks = new TreeSet<>(
     new PreUpgradeCheckComparator());
 
+  /**
+   * The upgrade checks discovered on a per-stack, per-upgrade pack basis.
+   */
+  private final Map<UpgradePack, Set<UpgradeCheck>> m_pluginChecks = new HashMap<>();
+
   /**
    * Register an upgrade check.
    *
    * @param upgradeCheck
    *          the check to register (not {@code null}).
    */
   public void register(UpgradeCheck upgradeCheck) {
-    m_upgradeChecks.add(upgradeCheck);
+    m_builtInChecks.add(upgradeCheck);
   }
 
   /**
-   * Gets an ordered list of all of the upgrade checks.
+   * Gets a list of all of the built-in upgrade checks.
    *
    * @return
    */
-  public List<UpgradeCheck> getUpgradeChecks() {
-    return new ArrayList<>(m_upgradeChecks);
+  public List<UpgradeCheck> getBuiltInUpgradeChecks() {
+    return new ArrayList<>(m_builtInChecks);
   }
 
-  public List<UpgradeCheck> getServiceLevelUpgradeChecks(UpgradePack upgradePack, Map<String, ServiceInfo> services) {
-    List<String> prerequisiteChecks = upgradePack.getPrerequisiteChecks();
-    List<String> missingChecks = new ArrayList<>();
-    for (String prerequisiteCheck : prerequisiteChecks) {
-      if (!isRegistered(prerequisiteCheck)) {
-        missingChecks.add(prerequisiteCheck);
+  /**
+   * Gets the ordered upgrade checks which should execute for the specified
+   * upgrade pack. This list of checks will include required built-in checks
+   * from Ambari and plugin checks loaded from the upgrade pack.
+   * <p>
+   * This collection of upgrade checks will be further scrutinized via
+   * {@link CheckQualification}s to determine if they indeed need to run by the
+   * {@link CheckHelper}.
+   *
+   * @param upgradePack
+   *          Upgrade pack object with the list of required checks to be
+   *          included
+   * @return
+   */
+  @SuppressWarnings("unchecked")
+  public List<UpgradeCheck> getFilteredUpgradeChecks(UpgradePack upgradePack) throws AmbariException {
```

Comment:
I would prefer this method to be splitted, since it has several distinct parts:

* filtering "required" built-in checks
* loading plugin checks via classloader
* caching plugin checks per upgrade pack
* combining built-in and plugin checks

Reply:
Yes, this method does several things, but it's still rather small. I could break out loading checks from the classloader if that would make the separation of concerns here clearer.